src/Rings/MPolyQuo.jl

```diff
@@ -101,6 +101,73 @@ function Base.deepcopy_internal(a::MPolyQuoRingElem, dict::IdDict)
   return MPolyQuoRingElem(Base.deepcopy_internal(a.f, dict), a.P, a.simplified)
 end
 
+########################################################################
+# Representatives of elements in quotient rings and normal forms 
+#
+# Elements [a] ∈ P/I in quotients of polynomial rings 
+# P = 𝕜[x₁,…,xₙ] by ideals I = ⟨f₁,…,fᵣ⟩ admit unique representatives 
+# a ∈ P whenever a normal form algorithm exists for the polynomial 
+# ring P. This is really a question about the ring of coefficients 𝕜. 
+#
+# In general, we can not expect a normal form algorithm to exist and, 
+# in particular, that it is implemented in Singular. However, we wish 
+# to use Singular as a default backend and this also drives the design 
+# of the quotient rings to begin with. 
+#
+# To make sure that the data structure for quotient rings can 
+# nevertheless also accomodate more exotic coefficient rings we 
+# provide the following functionality to decide the existence and use 
+# of a singular backend depending on the type.
+########################################################################
+
+# The trait of the coefficient ring and its elements to decide on 
+# whether or not we do have a singular backend for groebner basis 
+# computations
+abstract type HasGroebnerAlgorithmTrait end
+abstract type HasNormalFormTrait <: HasGroebnerAlgorithmTrait end
+
+# A normal form algorithm (for possibly non-global orderings) requires 
+# strictly more, so we make it a subcase.
+struct HasSingularNormalForm <: HasNormalFormTrait end
+struct HasNoNormalForm <: HasNormalFormTrait end
+
+struct HasSingularGroebnerAlgorithm <: HasGroebnerAlgorithmTrait end
+struct HasRingFlattening <: HasGroebnerAlgorithmTrait end
+struct HasNoGroebnerAlgorithm <: HasGroebnerAlgorithmTrait end
+
+# By default we do not expect the Singular backend to handle the case
+HasNormalFormTrait(a::Ring) = HasNormalFormTrait(typeof(a))
+HasNormalFormTrait(a::RingElem) = HasNormalFormTrait(parent_type(a))
+HasNormalFormTrait(::Type{T}) where {T <: RingElem} = HasNormalFormTrait(parent_type(T))
+HasNormalFormTrait(::Type{T}) where {T} = HasNoNormalForm()
+
+# Same for the Groebner bases. But if a normal form algorithm exists, then 
+# this works, too.
+HasGroebnerAlgorithmTrait(a::Ring) = HasGroebnerAlgorithmTrait(typeof(a))
+HasGroebnerAlgorithmTrait(a::RingElem) = HasGroebnerAlgorithmTrait(parent_type(a))
+HasGroebnerAlgorithmTrait(::Type{T}) where {T <: RingElem} = HasGroebnerAlgorithmTrait(parent_type(T))
+
+function HasGroebnerAlgorithmTrait(::Type{T}) where {T}
```

Comment:
Should this perhaps be
```suggestion
function HasGroebnerAlgorithmTrait(::Type{T}) where {T <: Ring}
```